63.CUDAInterop/main.cpp

```diff
@@ -0,0 +1,411 @@
+			re &= cmd->pipelineBarrier(EDF_NONE, { .bufBarrierCount = 1, .bufBarriers = &bufBarrier}); 
+
+			IGPUCommandBuffer::SBufferCopy region = { .size = size };
+			re &= cmd->copyBuffer(importedbuf.get(), stagingbuf.get(), 1, &region);
+
+			IGPUCommandBuffer::SImageMemoryBarrier<IGPUCommandBuffer::SOwnershipTransferBarrier> imgBarrier = {
+				.barrier = { 
+					.dep = { 
+						// .srcStageMask = PIPELINE_STAGE_FLAGS::ALL_COMMANDS_BITS,
+						// .srcAccessMask = ACCESS_FLAGS::MEMORY_READ_BITS | ACCESS_FLAGS::MEMORY_WRITE_BITS,
+						.dstStageMask = PIPELINE_STAGE_FLAGS::ALL_COMMANDS_BITS, 
+						.dstAccessMask = ACCESS_FLAGS::MEMORY_READ_BITS | ACCESS_FLAGS::MEMORY_WRITE_BITS,
+					},
+					.ownershipOp = IGPUCommandBuffer::SOwnershipTransferBarrier::OWNERSHIP_OP::ACQUIRE,
+					.otherQueueFamilyIndex = queue->getFamilyIndex(),
+				},
+				.image = importedimg.get(),
+				.subresourceRange = {
+					.aspectMask = IImage::EAF_COLOR_BIT,
+					.levelCount = 1u,
+					.layerCount = 1u,
+				},
+				.oldLayout = IImage::LAYOUT::PREINITIALIZED,
+				.newLayout = IImage::LAYOUT::TRANSFER_SRC_OPTIMAL,
+			};
+
+			re &= cmd->pipelineBarrier(EDF_NONE, {.imgBarrierCount = 1, .imgBarriers = &imgBarrier });
+
+			IImage::SBufferCopy imgRegion = {
+				.imageSubresource = {
+					.aspectMask = imgBarrier.subresourceRange.aspectMask,
+					.layerCount = imgBarrier.subresourceRange.layerCount,
+				},
+				.imageExtent = importedimg->getCreationParameters().extent,
+			};
+
+			re &= cmd->copyImageToBuffer(importedimg.get(), imgBarrier.newLayout, stagingbuf2.get(), 1, &imgRegion);
+			re &= cmd->end();
+		
+			IQueue::SSubmitInfo submitInfo = {
+				.waitSemaphores = std::array{IQueue::SSubmitInfo::SSemaphoreInfo{ .semaphore = sema.get(), .value = 0, .stageMask = PIPELINE_STAGE_FLAGS::ALL_COMMANDS_BITS, }},
+				.commandBuffers = std::array{IQueue::SSubmitInfo::SCommandBufferInfo{cmd.get()}},
+			};
+
+			auto submitRe = queue->submit(std::array{submitInfo});;
+			re &= IQueue::RESULT::SUCCESS == submitRe;
+			assert(re);
+		}
+        
+		ASSERT_SUCCESS(cu.pcuLaunchHostFunc(stream, [](void* userData) { decltype(this)(userData)->kernelCallback(); }, this));
+	}
+
+	void kernelCallback()
+	{
+		// Make sure we are also done with the readback
+		{
+			//IGPUFence* fences[] = { fence.get() };
+			//auto status = logicalDevice->waitForFences(1, fences, true, -1);
+			//assert(IGPUFence::ES_SUCCESS == status);
+		}
+
+		float* A = reinterpret_cast<float*>(cpubuffers[0]->getPointer());
+		float* B = reinterpret_cast<float*>(cpubuffers[1]->getPointer());
+		float* CBuf = reinterpret_cast<float*>(stagingbuf->getBoundMemory().memory->getMappedPointer());
+		float* CImg = reinterpret_cast<float*>(stagingbuf2->getBoundMemory().memory->getMappedPointer());
+
+		 assert(!memcmp(CBuf, CImg, size));
+
+		for (auto i = 0; i < numElements; i++)
+		{
+			assert(abs(CBuf[i] - A[i] - B[i]) < 0.01f);
+			assert(abs(CImg[i] - A[i] - B[i]) < 0.01f);
```

Comment:
`assert` compiles to noop in non-debug builds, seee ex 02, 03, 05 etc for more robust testing with logging